File: go/patterns/multiple_pointers_test.go
package patterns

import (
	"reflect"
	"testing"
)

func TestSumZero(t *testing.T) {
	t.Run("it should return nil if the array is empty", func(t *testing.T) {
		test := []int{}

		got := SumZero(test)

		assertNil(t, got)
	})

	t.Run("it should return the first pair of the array", func(t *testing.T) {
		test := []int{-8, -5, -3, 0, 1, 4, 8}

		got := SumZero(test)
		want := []int{-8, 8}

		if !reflect.DeepEqual(got, want) {
			t.Fatalf("got %v, want %v", got, want)
		}
	})

	t.Run("it should return nil when there is no pair where sum is zero", func(t *testing.T) {
		test := []int{1, 2, 3, 4, 5}

		got := SumZero(test)

		assertNil(t, got)
	})

	t.Run("it should return pair in the middle of the array", func(t *testing.T) {})
	test := []int{-8, -5, -3, 0, 3, 4, 9}

	got := SumZero(test)
	want := []int{-3, 3}

	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v, want %v", got, want)
	}
}

func assertNil(t *testing.T, got []int) {
	t.Helper()

	if got != nil {
		t.Fatalf("expected nil but received %v", got)
	}
}

File: javascript/src/patterns/multiple-pointers.test.ts
import { sumZero } from './multiple-pointers'

describe('Multiple Pointers Pattern', () => {
    it('should return undefined on empty array', () => {
        const got = sumZero([])
        expect(got).toBeUndefined()
    })

    it('should find the first pair where sum is zero', () => {
        const test = [-16, -2, -1, 0, 1, 2, 16]
        const got = sumZero(test)
        const want = [-16, 16]
        expect(got).toMatchObject(want)
    })

    it('should find in any location pair where sum is zero', () => {
        const test = [-16, -8, -2, 0, 2, 5, 17]
        const got = sumZero(test)
        const want = [-2, 2]
        expect(got).toMatchObject(want)
    })
})
